### PR TITLE
Add length validator

### PR DESCRIPTION
**Pull request by dhruvCW**
This validator validates that the length of the parameter (if the parameter supports `#length` is within the specified limits).

I've also added the `min_length` and `max_length` limits to the attribute documentation so that they can be used to specify the limits for an array type in swagger.

### CODE REVIEW

File: lib/grape/locale/en.yml [hunk 1]
```diff
@@ -10,6 +10,9 @@ en:
         values: 'does not have a valid value'
         except_values: 'has a value not allowed'
         same_as: 'is not the same as %{parameter}'
+        length_mismatch_both: 'is expected to have a length within %{min_length} and %{max_length}'
```
Comment (dblock) [inline lib/grape/locale/en.yml:13]: potentially better wordsmithing:

is expected to be between ... and ... characters long
is expected to be at least ... characters long
is expected to be less than ... characters long

Comment (dblock) [inline lib/grape/locale/en.yml:13]: let's match the variables of the other errors

```
length
length_min
length_max
```

File: lib/grape/validations/attributes_doc.rb [hunk 2]
```diff
@@ -28,6 +28,12 @@ def extract_details(validations)
         details[:documentation] = documentation if documentation
 
         details[:default] = validations[:default] if validations.key?(:default)
+
+        return unless validations.key?(:length)
```
Comment (dblock) [inline lib/grape/validations/attributes_doc.rb:32]: we'll be adding other validations, so let's not prematurely return, a double `if` is fine, it will get optimized away

```
details[:min_length] = validations[:length][:min] if validations.key?(:length) && validations[:length].key?(:min)
details[:max_length] = validations[:length][:max] if validations.key?(:length) && validations[:length].key?(:max)

```

File: lib/grape/validations/validators/length_validator.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Grape
+  module Validations
+    module Validators
+      class LengthValidator < Base
+        def initialize(attrs, options, required, scope, **opts)
+          @min_length = options[:min]
```
Comment (dblock) [inline lib/grape/validations/validators/length_validator.rb:8]: Call this `@min` and `@max` since we're already inside a `LengthValidator`

File: lib/grape/validations/validators/length_validator.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+          @max_length = options[:max]
+          super
+        end
+
+        def validate_param!(attr_name, params)
+          param = params[attr_name]
+          return unless param.respond_to?(:length)
+          return unless (@min_length && param.length < @min_length) || (@max_length && param.length > @max_length)
```
Comment (dblock) [inline lib/grape/validations/validators/length_validator.rb:16]: Semantically we probably don't want to treat `@min_length` of zero like this, so use `@min_length.nil?` in the condition?

**Review by dblock — CHANGES_REQUESTED** [review]
Looks good, some nits below, plus:

1. Add README documentation.
2. Add tests for zero.
3. Handle/add tests for min = max, min > max and min > max. Probably needs to error on some.

File: lib/grape/validations/validators/length_validator.rb (hunk 3 above)
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb]: done 👍

File: lib/grape/validations/validators/length_validator.rb (hunk 4 above)
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb]: makes sense done 👍

File: lib/grape/validations/attributes_doc.rb (hunk 2 above)
Comment (dhruvCW) [reply lib/grape/validations/attributes_doc.rb]: done 👍

File: lib/grape/locale/en.yml (hunk 1 above)
Comment (dhruvCW) [reply lib/grape/locale/en.yml]: done 👍

Comment (dhruvCW) [reply lib/grape/locale/en.yml]: I hesitate to use `characters long` since it doesn't make sense for arrays.

File: lib/grape/validations/validators/length_validator.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  module Validations
+    module Validators
+      class LengthValidator < Base
+        def initialize(attrs, options, required, scope, **opts)
+          @min = options[:min]
+          @max = options[:max]
+
+          super
+
+          raise ArgumentError, "min #{@min} cannot be greater than max #{@max}" if !@min.nil? && !@max.nil? && @min > @max
```
Comment (dblock) [inline lib/grape/validations/validators/length_validator.rb:13]: Should we handle negative values here too?

File: spec/grape/validations/validators/length_spec.rb [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        requires :list, type: [Integer], length: { max: 3 }
+      end
+      post 'with_max_only' do
+      end
+
+      params do
+        requires :list, type: Integer, length: { max: 3 }
+      end
+      post 'type_is_not_array' do
+      end
```
Comment (dblock) [inline spec/grape/validations/validators/length_spec.rb:28]: This is confusing. What is the length of an integer?

**Review by dblock — CHANGES_REQUESTED** [review]
I think this needs a bit more work around types. 

1. Length is not supported on types such as Integer, and should fail accordingly. Needs tests.
2. Semantically, `length` and `size` aren't the same thing, even though hashes for example support both length and size. Should we have interchangeable `length` and `size` validators?
3. What happens with comparable types, e.g. when I say `min: 'xyz'`?
4. Should this be a more generic validator that works on anything `Comparable`?

**Comment by dhruvCW** [comment]
> 1. Length is not supported on types such as Integer, and should fail accordingly. Needs tests.

makes sense should an argument error be raised in this case 🤔 

> 2. Semantically, `length` and `size` aren't the same thing, even though hashes for example support both length and size. Should we have interchangeable `length` and `size` validators?

While I agree, I see that in ruby both are interchangable so maybe an alias for length validator 🤔 

> 3. What happens with comparable types, e.g. when I say `min: 'xyz'`?
> 4. Should this be a more generic validator that works on anything `Comparable`?

IMO that won't make sense I can restrict the `min` and `max` values to be integer numbers (and maybe also that they not negative).

File: lib/grape/validations/attributes_doc.rb [hunk 7]
```diff
@@ -28,6 +28,9 @@ def extract_details(validations)
         details[:documentation] = documentation if documentation
 
         details[:default] = validations[:default] if validations.key?(:default)
+
+        details[:min_length] = validations[:length][:min] if validations.key?(:length) && validations[:length].key?(:min)
```
[the remark this answers is not in the document]
Comment (dhruvCW) [reply lib/grape/validations/attributes_doc.rb:32]: this results in a rubocop error that then suggests replacing the `if` with  an `unless` like I had originally written. How do you propose we proceed 🤔 

```ruby
return unless validations.key?(:length)

details[...]
details[...]
```

File: README.md [hunk 8]
```diff
@@ -1709,6 +1711,20 @@ params do
...
+
+Parameters can be restricted from having a specific length or size with the `:length` option.
+
+the validator accepts `:min` or `:max` or both options to validate that they parameter is within
+the given limits.
+
+```ruby
+params do
+  requires :str, length: { min: 3 }
+  requires :list, length: { min: 3, max: 5}
```
[the remark this answers is not in the document]
Comment (dblock) [reply README.md:1724]: Also format (missing space before }).

**Comment by dblock** [comment]
> IMO that won't make sense. I can restrict the `min` and `max` values to be integer numbers (and maybe also that they not negative).

Yes, I think that would be the correct behavior.

**Comment by dblock** [comment]
> > 1. Length is not supported on types such as Integer, and should fail accordingly. Needs tests.
> 
> makes sense should an argument error be raised in this case 🤔

Yes, I think anything being sent that doesn't have `.length` should yield a clear error. 
 
> > 2. Semantically, `length` and `size` aren't the same thing, even though hashes for example support both length and size. Should we have interchangeable `length` and `size` validators?
> 
> While I agree, I see that in ruby both are interchangable so maybe an alias for length validator 🤔

I think I'd like to write `length: ... ` which causes the validator to invoke `.length`, and `size: ` which causes the validator to invoke `.size`. I would implement both validators with a base "comparable" class that we can reuse for a future (or part of this PR maybe even) "compare: { min: max: }" validator that would work for strings too?

File: README.md [hunk 9]
```diff
@@ -1709,6 +1711,20 @@ params do
 end
 ```
 
+#### `length`
+
+Parameters can be restricted from having a specific length or size with the `:length` option.
```
[the remark this answers is not in the document]
Comment (dhruvCW) [reply README.md]: done restricted the validator to only string or array and mentioned that in the readme as well.

File: lib/grape/validations/validators/length_validator.rb (hunk 5 above)
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb:15]: done restricted allowed values to non negative integer numbers.

File: spec/grape/validations/validators/length_spec.rb (hunk 6 above)
Comment (dhruvCW) [reply spec/grape/validations/validators/length_spec.rb:28]: yup now if the type is not string or array it will raise an `ArgumentError`.

**Comment by dhruvCW** [comment]
> > > 1. Length is not supported on types such as Integer, and should fail accordingly. Needs tests.
> > 
> > 
> > makes sense should an argument error be raised in this case 🤔
> 
> Yes, I think anything being sent that doesn't have `.length` should yield a clear error.
> 
> > > 2. Semantically, `length` and `size` aren't the same thing, even though hashes for example support both length and size. Should we have interchangeable `length` and `size` validators?
> > 
> > 
> > While I agree, I see that in ruby both are interchangable so maybe an alias for length validator 🤔
> 
> I think I'd like to write `length: ... ` which causes the validator to invoke `.length`, and `size: ` which causes the validator to invoke `.size`. I would implement both validators with a base "comparable" class that we can reuse for a future (or part of this PR maybe even) "compare: { min: max: }" validator that would work for strings too?

I have restricted the validator to only validate `String` or `Array` types. I am happy to alias the validator to support `size: ` as well but I think for this PR I want to be restrictive and only support these two concrete types instead of any class that has a `length` or `size` method.

File: lib/grape/validations/attributes_doc.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (dblock) [reply lib/grape/validations/attributes_doc.rb:32]: We can leave it.

File: lib/grape/validations/validators/length_validator.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+          raise ArgumentError, 'min must be an integer greater than or equal to zero' if !@min.nil? && (!@min.is_a?(Integer) || @min.negative?)
+          raise ArgumentError, 'max must be an integer greater than or equal to zero' if !@max.nil? && (!@max.is_a?(Integer) || @max.negative?)
+          raise ArgumentError, "min #{@min} cannot be greater than max #{@max}" if !@min.nil? && !@max.nil? && @min > @max
+        end
+
+        def validate_param!(attr_name, params)
+          param = params[attr_name]
+          param = param.compact if param.respond_to?(:compact)
+
+          raise ArgumentError, "parameter #{param} has an unsupported type. Only strings & arrays are supported" unless params.is_a?(String) || param.is_a?(Array)
```
Comment (dblock) [inline lib/grape/validations/validators/length_validator.rb:22]: Should this be anything that `.respond_to?(:length)`? 

nit: Let's spell out `&` as `and`

**Review by dblock — COMMENTED** [review]
We can leave the size/length for a future PR. But I am not sure restricting to Array and String is correct.

File: lib/grape/validations/validators/length_validator.rb (hunk 10 above)
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb:22]: Nope I decided on only supporting `String` and `Array` for now since it doesn't make sense to me to support other types e.g. `Hash`.

**Comment by dblock** [comment]
I noticed the `compact`. That forces an interesting behavior of removing `nil` entries, which I don't think is right, `nil`s are very valid values too, especially in content types like JSON, so let's remove it (compact can be introduced as a future option).

**Comment by dhruvCW** [comment]
> I noticed the `compact`. That forces an interesting behavior of removing `nil` entries, which I don't think is right, `nil`s are very valid values too, especially in content types like JSON, so let's remove it (compact can be introduced as a future option). Add tests with a `nil` value that would fail with a `.compact`.

the problem is when an empty array is passed as the parameter the validator receives `[nil]`. How do you propose we handle this situation 🤔

**Comment by dblock** [comment]
> I have restricted the validator to only validate `String` or `Array` types. I am happy to alias the validator to support `size: ` as well but I think for this PR I want to be restrictive and only support these two concrete types instead of any class that has a `length` or `size` method.

Why? I think `Hash` is an important one to support for a length validator, as an example.

**Comment by dhruvCW** [comment]
> But I am not sure restricting to Array and String is correct.

could you maybe elaborate on your concern 🤔 . 

My line of thinking is that a type with a variable "length" (number of contained elements) like `String` or `Array` should have such a validation but for a type like `Hash` which is usually well specified (known set of keys) it doesn't make sense to validate the "length" or "size" of the object.

**Comment by dblock** [comment]
> > I noticed the `compact`. That forces an interesting behavior of removing `nil` entries, which I don't think is right, `nil`s are very valid values too, especially in content types like JSON, so let's remove it (compact can be introduced as a future option). Add tests with a `nil` value that would fail with a `.compact`.
> 
> the problem is when an empty array is passed as the parameter the validator receives `[nil]`. How do you propose we handle this situation 🤔

That is a side effect of rack-test or rack parser that we don't control if I am not mistaken. There's a bunch of issues filed along those lines (e.g. https://github.com/ruby-grape/grape/issues/1370). For the tests for arrays switch to sending JSON which will match exactly what you're sending server-side.

**Comment by dblock** [comment]
> could you maybe elaborate on your concern 🤔 .
> 
> My line of thinking is that a type with a variable "length" (number of contained elements) like `String` or `Array` should have such a validation but for a type like `Hash` which is usually well specified (known set of keys) it doesn't make sense to validate the "length" or "size" of the object.

Take a typical rollup API that receives some stats for a given, say SKU. 

```json
{
   'sku1' : {
      total: 123.
      name: 'sku one'
   },
   'sku2' : {
      total: 234.
      name: 'sku two'
   }
}
```

I choose a hash because I don't want sku's to repeat themselves, and with the length validator I'd be able to limit how many `sku`'s one could send at any given time. 

The other concern is that if we restruct `length` on certain types it will be part of the API forever. Relaxing it would require a backwards-incompatible change. If we make it `respond_to? :length`, we would defer to the broadest range of capabilities (and allow people to shoot themselves in the foot, which is fine for an API library).  

WDYT?

**Comment by dblock** [comment]
Btw, thank you for hanging in here with me! Really appreciate your time and work on this.

**Comment by dhruvCW** [comment]
> The other concern is that if we restrict `length` on certain types it will be part of the API forever. Relaxing it would require a backwards-incompatible change. If we make it `respond_to? :length`, we would defer to the broadest range of capabilities (and allow people to shoot themselves in the foot, which is fine for an API library).
> 
> WDYT?

Ah makes sense ✅ 

>and allow people to shoot themselves in the foot, which is fine for an API library

😅 wasn't sure if this is the library stance or not 🙈

**Comment by dhruvCW** [comment]
> > > I noticed the `compact`. That forces an interesting behavior of removing `nil` entries, which I don't think is right, `nil`s are very valid values too, especially in content types like JSON, so let's remove it (compact can be introduced as a future option). Add tests with a `nil` value that would fail with a `.compact`.
> > 
> > 
> > the problem is when an empty array is passed as the parameter the validator receives `[nil]`. How do you propose we handle this situation 🤔
> 
> Either way, if a client sends an empty array, the server is expected to receive an empty array (`[nil]` is not an empty array). I believe that is a side effect of rack-test or rack parser that we don't control for GETs or POSTs with encoded form content type or something like that. There's a bunch of issues filed along those lines (e.g. #1370). For the tests for arrays switch to sending JSON which will match exactly what you're sending server-side.

just tried to change the tests to match the type to `[JSON]` instead of `[Integer]` and now when passing an empty array it fails in the coercer (as an invalid type).

File: lib/grape/validations/validators/length_validator.rb (hunk 10 above)
Comment (dhruvCW) [reply lib/grape/validations/validators/length_validator.rb]: removed the restriction and now allows any type that responds to `#length`.

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
Merged, thank you!

Want to try to implement `size` without duplicating all the code on top of this?

**Comment by dhruvCW** [comment]
> Merged, thank you!
> 
> Want to try to implement `size` without duplicating all the code on top of this?

😅 For now no. I'll try the next time I have some time to spare.

**Comment by dhruvCW** [comment]
@dblock will you be releasing 2.1.0 soon ?

**Comment by dblock** [comment]
@dhruvCW I'll take are of it this week
